### ISSUE TITLE
Signed version of SQLitePCL.bundle_green 

### ISSUE THREAD

**Issue by pragnagopa** [ericsink/SQLitePCL.raw#93]
We have recently updated https://github.com/Azure/azure-mobile-apps-net-client to take dependency on SQLitePCL.bundle_green 0.9.3. Loading SQLitePCL.raw fails at runtime:
Could not load file or assembly 'SQLitePCL.raw, Version=0.9.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. A strongly-named assembly is required. (Exception from HRESULT: 0x80131044)

All the assemblies in [Microsoft.Azure.Mobile.Client](https://www.nuget.org/packages/Microsoft.Azure.Mobile.Client/ ) are signed. Is it possible to publish a nuget package with signed assemblies?

**Comment by ericsink** [comment]
For context, this looks like a new form of issue #53.  Same reply:  Yes, I will make this happen, and quickly as I can, but strong-naming has been a can of worms.  I need to re-open the can, remember the issues, and push through them.

**Comment by pragnagopa** [comment]
@ericsink Thanks for the quick follow up!
FYI: @paulbatum,  @fabiocav

**Comment by ericsink** [comment]
If I am looking at the correct packages on nuget, it looks like you have not yet pushed packages with the SQLitePCL.raw dep.  Is that right?

In other words, the urgency on the strong-naming issue for you may be "high", but it is not "urgent crisis because everything is broken".  Is that right?

**Comment by pragnagopa** [comment]
We need to release ASAP as our current SDK will not work with Android N. Our current release is blocked until Strong Name issue is fixed. Let me know if there is anything I can help with.

**Comment by ericsink** [comment]
Understood.  And also:

I have been working on my 0.9.4 release, which contains a pretty significant number of changes since 0.9.3.  I assume those changes represent risk to you, and you would prefer something closer to just "0.9.3 + strong naming"?

**Comment by pragnagopa** [comment]
Let me test our bits with 0.9.4-pre1 nuget package as we would prefer depending on latest nuget packages.

**Comment by ericsink** [comment]
Er, actually...

0.9.4-pre1 does not contain most of the changes that I've been planning to be in 0.9.4.  So that testing won't help you.

Roughly speaking, 0.9.4 contains:

fix for certain cases where UWP apps fail validation

support for custom sqlite3 builds

support for using the winsqlite3 built-in to Windows 10

refactor and rename some things that where the need to deal with Android N has broken old assumptions

Bottom line:  I would like you guys to end up on 0.9.4.  I'll try to get at least a pre-release build pushed as soon as possible so you can be heading in that direction.

**Comment by pragnagopa** [comment]
Ok. Thanks.  0.9.4 + Strong Name should work for us. I can work on updating out SDK if there are any breaking changes.

**Comment by ericsink** [comment]
I have two options here:

1.  Become confident that I can add strong naming for all SQLitePCL.raw users without causing difficulties for those who do not need it.

2.  Begin publishing two parallel sets of packages, one with strong naming (for those who need it), and one without.

I do not know how much time option 1 will take.  

Option 2 is relatively easy.  My build system is entirely build around automatic generation of code and project files, so the SN packages and the non-SN packages will always be identical except for the strong naming.

How do you feel about option 2?

**Comment by pragnagopa** [comment]
Yes. Option 2 sounds good. Thanks!

**Comment by ericsink** [comment]
I've implemented the changes to the build system to create two parallel sets of packages.

The ones starting with "SQLitePCL." are the regular ones.

The ones starting with "SQLitePCL_sn." are the ones with strong names.

After some more testing and fixes, I will push a prerelease of 0.9.4, hopefully later today or tomorrow.

**Comment by pragnagopa** [comment]
Ok. Thanks for the update.

**Comment by jamesmontemagno** [comment]
Could the Azure SDK just use: https://github.com/dsplaisted/strongnamer

**Comment by ericsink** [comment]
I'll let @pragnagopa say whether the @dplaisted strongnamer tool would meet their requirements or not.

But if it would, I would be delighted to remove all the SN stuff I've added to my build system in the last couple days.  :-)

**Comment by pragnagopa** [comment]
Will look into it and get back to you soon.

**Comment by pragnagopa** [comment]
Discussed with my team. Looks like we have already looked into strongnamer but are reluctant to add another dependency as this would force our customers to take a dependency on it as well.

**Comment by ericsink** [comment]
But if the technique works...

Where does the extra dep become a problem?

What if I had implemented compatibility with strongnames by simply adding that dep to bundle_green?

What if I extra the concept or technique from strongnamer and put it into a bundle_green_sn?

I intend to do whatever is needed to make sure my lib is compatible with your situation, but I'd rather add 1 new package than 56.

**Comment by pragnagopa** [comment]
Thank you for raising your concerns.
Where does the extra dep become a problem?
1. Once we package Azure Mobile Apps nuget with strongnamer, it will be part of customer's application and will affect all the assemblies that are part of the application as we cannot restrict strong naming to just the assemblies we depend on.
2.  For any additional dependency we add, we will have an added maintenance cost as we need to thoroughly test before any release/upgrade to make sure we don't break our customers.

What if I had implemented compatibility with strongnames by simply adding that dep to bundle_green?
If bundle_green adds strongnamer and if that means all the assemblies in the application will be affected, then we will have to re-evaluate our dependency bundle_green as it will require a security review.

What if I extra the concept or technique from strongnamer and put it into a bundle_green_sn?
If bundle_green_sn uses the same technique as strongnamer, this would be fine assuming only assemblies in bundle_green_sn will be affected.

Hope I answered all your questions!

**Comment by dsplaisted** [comment]
This is a thorny issue. Please consider discussing it with @terrajobst and @davkean from the .NET team (and feel free to include me too if you can wait until Thursday or later when I’m back from vacation).
 
Having separate signed and unsigned versions of SQLitePCL.raw is highly undesirable, as it means that all libraries that depend on SQLitePCL.raw would also have to have separate signed and unsigned versions, otherwise they would be incompatible with each other.  For example, if [sqlite-net-pcl](https://www.nuget.org/packages/sqlite-net-pcl/) only has an unsigned version, and Microsoft.Azure.Mobile.Client only has a signed version, then it won't be possible to use both of the packages together in an app.

There's not too much of a downside to strong naming an assembly, so it would probably be OK to just switch over to strong naming SQLitePCL.raw for everyone.  However, adding (or removing) a strong name is a breaking change that would require all existing libraries (and NuGet packages) that depend on it to be recompiled with the newly strong named version.

There's also not a strong reason for a library to be strong named.  Even for a library shipped by Microsoft, I think we could consider whether it's actually the right thing to do to strong name it.  However, since removing a strong name is a breaking change, this may not be an option for the Microsoft.Azure.Mobile.Client package.

Another option might be to use something like strong namer to sign and then ILMerge SQLitePCL.raw into Microsoft.Azure.Mobile.Client.  That would mean instead of having a package dependency on it, that the Azure library would include a copy of the SQL library merged into its own assembly.

> If bundle_green adds strongnamer and if that means all the assemblies in the application will be affected, then we will have to re-evaluate our dependency bundle_green as it will require a security review.

@pragnagopa Strong naming should not be used for security, so your statement makes me slightly worried that strong naming is being used incorrectly somewhere.

**Comment by pragnagopa** [comment]
Sorry! I just found that we have a requirement on assemblies being strongly named at compile time as we are part of visual studio. So including strongnamer will not work for us.
@dsplaisted, including sqlpcl.raw libraries directly in our package will restrict customers from upgrading to any other version of bundle-green

**Comment by ericsink** [comment]
Yep, issues don't come much thornier than this one.  I've taken a lap around this track once before, about a year ago, in issue #53.

> There's not too much of a downside to strong naming an assembly, 

The prevailing sentiment in the OSS community would seem to disagree with this.  Vehemently.  But I can't really know how much the pain is being exaggerated, because I haven't experienced it myself.

I've also heard lots from some in the "strongnames are the best thing ever" camp.  The exaggeration from that side comes in the form of denying that tradeoffs exist.  (There are no problems with strongnaming.  The coin has only one side.)  I haven't found those arguments helpful.

I consider it **essential** that I find a way to make sure that people who need to use strong names (like @pragnagopa and his team) can use SQLitePCL.raw.

But forcing strong names on all my users seems unthinkable, given (1) the huge tide of opposing sentiment, and (2) my lack of experiential knowledge on how to minimize the pain.

So I had really convinced myself that having a parallel set of packages would be the least bad option.  Still dreadful, but less so than anything else.

But I do want to explore the idea of building one of my "bundle" packages which simply brings in the deps it needs and uses the same technique as strongnamer to present them as strongnamed assemblies even though they are not.  If I understand the concern from @pragnagopa it is simply that he wants the black magic confined to the SQLitePCL.raw packages, with no chance of other assemblies getting affected.  In other words, the scary thing about strongnamer is that you drop it in and it (apparently) just looks at all the references and possible does its magic on all of them, whereas he and his team want its possible affects to be strictly limited.  Correct me if I'm wrong.

**Comment by pragnagopa** [comment]
I just found that we have a requirement on assemblies being strongly named at compile time as we are part of visual studio Which requires assemblies to be strongly named at compile time.

**Comment by ericsink** [comment]
FWIW, here is part of how I talked myself into a parallel set of packages:

SQLitePCL.raw is open source.  Apache license.  Anybody can fork it and do what they want.

If somebody doesn't like the lack of strong names on the assemblies, they can fork SQLitePCL.raw, add strong names to everything, and re-release it.

Of course, they'll have to use separate package names.  So their forked version will be a completely different thing, with no expectation of compatibility or interop with SQLitePCL.raw.  People who need strong names would choose that other package.  Everybody else would choose SQLitePCL.raw.

Now what if that "somebody" was me?

**Comment by pragnagopa** [comment]
One of the reasons we decided to take dependency on bundle_green instead of updating our current dependency: SQLitePCL to work with Android N, is that it simplifies our  build process and improves customer experience. Thanks @ericsink for considering our requirements!

**Comment by ericsink** [comment]
I should also acknowledge that option 1 is still on the table.  From above, option 1:

> Become confident that I can add strong naming for all SQLitePCL.raw users without causing difficulties for those who do not need it.

This is the path where I figure out exactly what @dsplaisted means when he says "There's not too much of a downside to strong naming an assembly".  Maybe strong names aren't that bad.  Maybe all the pain is exaggerated.  (Like I said, I wouldn't know.  I left the .NET community for a few years to go do something stupid, and now that I'm back, it looks like I missed a few things.)

But this path would probably take time.  And @pragnagopa is in a hurry, since Android N is shipping, and dealing with that mess is apparently part of their reason for taking a dep on SQLitePCL.raw.

So, I want to find the right answer here, but I also want to find it within the time constraints.

Anyway, perhaps I should stop adding comments on this item and go back to trying to find the right answer.

**Comment by ericsink** [comment]
OK, I am trying to talk myself into just adding strong names to all SQLitePCL.raw assemblies.  I have gone back and re-read a bunch of old stuff on the topic, including the JSON.NET war, and the octokit war, and so on.

My first high-level observation is that the strong name wars of 2012-2014 happened in a different context, one where the .NET open source community was mostly distinct from Microsoft.  That is no longer true.

For me, all the arguments about the benefits of strong naming as a feature are irrelevant.  The only benefit of strong naming that matters to me is that projects like Microsoft.Azure.Mobile.Client will be able to use my library.  To make that happen, I need to either strong name or I need to start convincing teams in Redmond to change their policy.  The latter path is absurd, so I *will* release a strong named version of SQLitePCL.raw.  The only question is whether it will be the only version.  Will I have one set of strong named packages, or two parallel sets, one with strong names and one without?

As I said above, to have one set, I need to:

> Become confident that I can add strong naming for all SQLitePCL.raw users without causing difficulties for those who do not need it.

AFAICT, most of the pain of using strong names can be removed by never changing the assembly version again.  This would apparently mean that using SQLitePCL.raw with the GAC could be problematic, but I don't think anyone does that anyway.  And I also think that nobody uses SQLitePCL.raw assemblies apart from the nuget packages, so the notion of abandoning the lower-level versioning scheme in favor of nuget's versioning scheme is viable in my case.

Another common argument is that strong naming an open source package makes it harder for new contributors to get started.  That's a non-issue for SQLitePCL.raw as well.  Basically, if I were trying to make it easy for people to contribute to this project, strong naming would be the least of the problems.

I am a little bit concerned about whether problems are going to occur with oddball platforms.  Did the previous strong naming wars happen mostly in the context of the .NET platform?  I support stuff like net35, Windows Phone Silverlight, netstandard, UWP, Xamarin, and so on.  And in every case, the primary function of this library is to facilitate calling native code (sqlite3 itself).  Lots of others have walked the "strong naming is stupid but fine I'll do it anyway" path, but I'm not sure any of those people have my particular set of problems.

But again, I only have two choices:  add strong names, or maintain my own strong-named fork.

As suggested by @dsplaisted above, I specifically invite feedback from @davkean and @terrajobst.

And also, since @davidfowl is an example of a notable Microsoft person who has taken the "just stop strong naming" stance in the past, I am curious about his current opinions on the matter.  

Thanks to all for your patience with this "thorny issue".

**Comment by AArnott** [comment]
@dsplaisted said:

> However, adding (or removing) a strong name is a breaking change that would require all existing libraries (and NuGet packages) that depend on it to be recompiled with the newly strong named version.

@terrajobst has previously said that _adding_ strong name signing to an assembly is not a breaking change. A binding redirect in the app can make it work. He has also indicated that _removing_ a strong name signature _is_ a breaking change as a binding redirect will not make that work.

@pragnagopa said:

> I just found that we have a requirement on assemblies being strongly named at compile time as we are part of visual studio Which requires assemblies to be strongly named at compile time.

I don't think that's true. I used to think it was true, but it turns out only authenticode signing is required. Strong-name signing is not required to ship with VS. However, not strong-name signing in VS may be very problematic in practice if you expect anyone else to reference you, since almost everyone else is strong-name signed. But what you can do is strong-name sign your contract assemblies and avoid a reference to sqlitePCL in that contract assembly. Then do _not_ strong name sign your implementation assembly and have that be what references SqlitePCL. That's how we've made do with sqlitePCL.raw so far (which already is in VS, BTW). 

That said, I'm fully in favor of this project starting to strong-name sign its assemblies. I'm just volunteering that sometimes there's an acceptable workaround.

**Comment by davkean** [comment]
Adding a strong name __is a__ breaking change - binding redirects will not help you here.

**Comment by AArnott** [comment]
Good to know. Thanks.

**Comment by pragnagopa** [comment]
Azure Mobile Apps is included as part of Visual Studio install and is used for connecting to Azure Mobile App runtime. Having a strong named assemblies is a requirement for us. My team and I greatly appreciate all the effort and time spent by @ericsink  and others to figure out the right solution. Unfortunately we are time constrained and we need to release ASAP. 

@ericsink assuming you will be taking the route of maintaining signed and unsigned versions as adding SN is a breaking change,  can you please give me a tentative timeline on when can you can publish bundle_green_sn?

**Comment by ericsink** [comment]
Hoping to publish a prerelease today (for your testing and integration purposes) and a final by the end of the week.

**Comment by ericsink** [comment]
@pragnagopa Could I provide you those prerelease testing packages out-of-band instead of publishing them through nuget.org?

**Comment by pragnagopa** [comment]
Yes. That will help to get started on the testing the new version. Please send them to pgopa@microsoft.com. It would be great if you can also publish prerelease to nuget once I am done testing so that we can release a prerelease version of Azure Mobile Apps while we wait for the final package.

**Comment by ericsink** [comment]
OK, my current plan is:

1.  One set of assemblies, with strong names
2.  But this next release will be version 1.0, not 0.9.4.  The AssemblyVersion will be 1.0.0.0 and will (probably) never change.
3.  And, in this release, all packages have names starting with SQLitePCLRaw.* instead of SQLitePCL.*

Background:  Actually this release had more breaking changes than just strong names.  Calling it 0.9.4 was always kind of a bad idea.  If I had pushed this as SQLitePCL.raw-0.9.4, and if Couchbase Lite users updated, things for them would break.  Bad idea.  So, giving everything a completely new name is more appropriate, and allows me to help migrate 0.9 users on whatever schedule makes sense for them.  

And given that everything is new, this is the time to add strong names and support them going forward.

A zip file with all this has been sent to @pragnagopa.  Once I get basic feedback on that, I'll be pushing the prerelease packages to nuget.org, proceeding to 1.0 final as quickly as prudence permits.

**Comment by dsplaisted** [comment]
> But this next release will be version 1.0, not 0.9.4. The AssemblyVersion will be 1.0.0.0 and will (probably) never change.

I understand why people go for this approach, but I don't recommend it.  That's because although strong naming generally doesn't provide any technical advantage, in the cases where there is a legitimate technical reason to use strong naming, the strategy of not changing the AssemblyVersion defeats the reason you had to use strong naming in the first place.

The main technical reasons to use strong names are:

- You need to put an assembly in the GAC, for whatever reason.
- You support third-party extensions.

The GAC is machine-wide and .NET will prefer to load an assembly from the GAC over an assembly with the same version in the app folder.  So if one app installs the library in the GAC, and another app depends on a later version of the library that has the same AssemblyVersion, then the second app will load the previous version from the GAC, and will probably fail with `MissingMethodException` or similar.

In an app (for example, Visual Studio) that loads third-party extensions into the same AppDomain, it's a good idea for reusable libraries used by extensions to be strong named.  This lets different versions of a reusable library be loaded side-by-side.  If two extensions depend on different versions of a library that is not strong named, then the version of the library loaded will be from whichever extension happens to load first.  This means that the other extension is likely to get `MissingMethodExceptions` or similar.  The same behavior (first extension wins) also occurs if different versions of the library use the same `AssemblyVersion`.

So in both cases where you actually need strong naming, releasing different versions of a library with the same `AssemblyVersion` can result in unexpected errors happening as a result of an unrelated app or extension being installed on the machine.

The reason you would want to leave the `AssemblyVersion` constant is to avoid binding redirects.  However, I think [Automatic Binding Redirects](https://msdn.microsoft.com/en-us/library/2fc472t2(v=vs.110).aspx) should make them a lot less painful than everyone remembers.

**Comment by terrajobst** [comment]
@AArnott 

> @terrajobst has previously said that adding strong name signing to an assembly is not a breaking change

As @davkean said: unfortunately that's not how it works.

**Comment by ericsink** [comment]
Thanks to @dsplaisted for the comment starting with "I understand why people go for this approach, but I don't recommend it."  That's more to think about.

Going forward, I simply want to navigate this issue in whatever fashion causes the least pain.  I'll stay open minded about exactly how that is done.

**Comment by ericsink** [comment]
Version 1.0 is released, with signed assemblies.
